[Jnjy] iP

Jnjy opened the pull request:
# Chad

>“Your mind is for having ideas, not holding them.” – David Allen (source) 

Chad, never worry about missing your deadlines anymore. Ths solution lies in me. It's
- text-based
- ~~easy~~ EFFORTLESS to learn
- ~~fast~~ **_SUPER_** FAST to use 

<br/>

All you need todo is,
1. download it from [here](https://github.com/Jnjy/ip/releases/tag/v0.1)
2. double-click it
3. **Chad** with me
4. let me manage your tasks for you :fire:

<br/>

You're a Brokie? No worries, it is **FREE**!

<br/>

Features:
- [x] Managing tasks (Mark, Unmark, Delete)
- [x] Managing deadlines
- [ ] Search tasks
- [ ] Insane Graphic User Interface (Coming soon)

<br/>
If you a true Java programmer, you can use it to practice Java too. Here's the main method:
<br/>
<br/>

``` java
public static void main(String[] args) {
    new Duke().run();
}
```

<br/>

Commit:
Add Gradle support

Jnjy committed:
Level-01: Added Greet, Echo and Exit method

Jnjy committed:
Level-01: Added printRes function to standardise printing format

Jnjy committed:
Level-01: Added javadocs

Jnjy committed:
Level-02: Added Add and List method

Jnjy committed:
Level-03: Added mark and unmark function

Jnjy committed:
Level-03: Added Request Class to handle the user's request

Jnjy committed:
Level-03: Added javadocs

Jnjy committed:
Level-04: Added ToDo, Event and Deadline

Jnjy committed:
A-TextUiTesting: Added input text and expected result

Jnjy committed:
Level-05: Handled Exception, edited code to incorporate OOP

Jnjy committed:
Level-06: Added delete command

Jnjy committed:
Level-06: Added IndexOutofBoundException

Jnjy committed:
Level-06: Minor update on exception

Jnjy committed:
A-TextUiTesting: Added new test cases for latest functions

Jnjy committed:
Level-07: edited tasklists to add tasks from file

Jnjy committed:
Level-07: added save function to save tasklist into file

Jnjy committed:
Added Dates formatting

Jnjy committed:
Merge branch 'branch-Level-7'

Jnjy committed:
Level-8: Merge + resolved conflicts

Jnjy committed:
A-MoreOOP: changed duke to become a class

Jnjy committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Jnjy committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Jnjy committed:
A-Gradle: set up Gradle

Jnjy committed:
Refactor: Request -> Parse and minor edit on code style

Jnjy committed:
A-JUnit: added several JUnit test cases

Jnjy committed:
A-Jar: adde Jar file

Jnjy committed:
A-JavaDoc: added JavaDoc for remaining files

Jnjy committed:
A-CodingStandard: Edited code to adhere to coding standard

Jnjy committed:
A-JavaDoc: added javadoc for class

Jnjy committed:
Level-09: Added find method

Jnjy committed:
Merge branch 'branch-A-JavaDoc'

Jnjy committed:
Merge: merge branch-A-CodingStandard and branch-A-JavaDoc after resolving conflicts

Jnjy committed:
Merge: Merge branch-Level-9 into master

Jnjy committed:
Edit: minor fix on code to adhere to coding standard

Jnjy committed:
Level-10: Added Hello World GUI (Basic) to Duke

Jnjy committed:
A-Packages: Re-organise packages

Jnjy committed:
Level-10: first iteration of java fx using java code

Jnjy committed:
Refactor: Minor refactor on Task class and clean the code

Jnjy committed:
Level-10: Change gui into using FXML

Jnjy committed:
Level-10: Integrate with Duke's task list

Jnjy committed:
Level-10: Added javadocs

Jnjy committed:
Merge branch 'branch-Level-10'

Jnjy committed:
Update README.md

Jnjy committed:
Update README.md

ScorpiusSigma commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,96 @@
+        } catch (DukeException e) {
+            return e.getMessage();
+        }
+    }
+    /**
```
Do you reckon there should be a new line in between the lines to separate the methods?

ScorpiusSigma commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,96 @@
+        UI.exit();
+    }
+
+    public String getResponse(String request) {
+        try {
+            Command com = new Parser(request, tasks).processRequest();
+            String response = com.execute(tasks);
+            return response;
+        } catch (DukeException e) {
+            return e.getMessage();
+        }
+    }
+    /**
+     * Function to handle the user's request
+     */
+    public static void handleRequest() {
+
+        Scanner sc = new Scanner(System.in);
+        String input = sc.nextLine();
+
+        TaskList tasks = new TaskList();
+
+        while (!input.equalsIgnoreCase("bye")) {
+            Command com = new Parser(input, tasks).processRequest();
+            String response = com.execute(tasks);
+            UI.printRes(response);
+            input = sc.nextLine();
+        }
+
+        sc.close();
+    }
```
Do you reckon there should be a gap between 2 methods between line 72 to

ScorpiusSigma commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,59 @@
+
+        String[] req = request.split("deadline ");
+
+        // check presence of task description
+        if (req.length < 2) {
+            throw new MissingArgumentException("☹ OOPS!!! You're missing the task description");
+        }
+
+        req = req[1].split("/by ");
+
+        // check presence of argument
+        if (req.length < 2) {
+            throw new MissingArgumentException("☹ OOPS!!! You're missing the task deadline");
+        }
```
How do you think this can be improved?

ScorpiusSigma commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,81 @@
+            throw new MissingArgumentException("Missing task description!");
+        } else if (req.length < 2) {
+            throw new MissingArgumentException("Please insert an start date.");
+        }
+
+        String[] duration = req[1].split(" /to ");
+
+        // check missing `end date`
+        if (duration.length < 2) {
+            throw new MissingArgumentException("Please insert an end date.");
+        }
+
+        String from = duration[0].strip();
+        String to = duration[1].strip();
+
+        // check missing `start date` and ` end date`
+        if (from.equals("")) {
+            throw new MissingArgumentException("Please insert a start date.");
+        } else if (duration.length < 2 || to.equals("")) {
+            throw new MissingArgumentException("Please insert an end date.");
+        }
+
+        try {
+            LocalDate startDate = LocalDate.parse(from);
+            LocalDate endDate = LocalDate.parse(to);
+            Event newEvent = tasks.addEvent(task, startDate, endDate);
+
+            // check valid `duration`
+            if (startDate.isAfter(endDate)) {
+                throw new InvalidArgumentException("Your start date should be before your end date!");
+            }
+
+            return "Great! I've added this task for you \n" + newEvent
+                    + "\nYou have " + tasks.numOfTask() + " tasks in the list";
+        } catch (DateTimeException error) {
+            throw new InvalidArgumentException("Wrong date format! "
+                    + "Please follow the format YYYY-MM-DD (e.g. 2000-01-01)");
+        }
+    }
+}
```
I'm not sure I like how long this is. Perhaps you can split the exception check to be in another method?

ScorpiusSigma commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public FindCommand(String request) {
+        this.request = request;
+    }
+
+    @Override
+    public String execute(TaskList tasks) {
```
Do you reckon adding javadocs to all your execute methods would allow the user to make sense of what that method does for that class?  This can be applied to other execute methods in other classes :)

securespider replied:
I agree i think doing so makes it more explicit and more clear for developers to understand what the execute does for each of the commands

ScorpiusSigma commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,18 @@
+public class ListCommand extends Command {
+    @Override
+    public String execute(TaskList tasks) {
+        // if there is no task in the list
```
I like how you have have inline comments which makes reading even better!

ScorpiusSigma reviewed (commented):
Overall, I like how you document your code and implementation. However, I felt some methods can be abstracted even further such as having another method to check for errors. Good Job! 👍

securespider commented on src/main/java/duke/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,37 @@
+
+        // check missing description of todo task
+        if (task.length == 0) {
+            throw new MissingArgumentException("☹ OOPS!!! The description of a todo cannot be empty.");
```
Instead of having this message as parameter in your constructor, if this is the default string for this exception it might be better to put it inside the exception itself

securespider commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,70 @@
+import duke.command.ListCommand;
+import duke.command.MarkCommand;
+import duke.command.RemoveCommand;
+import duke.command.UnmarkCommand;
```
Good job! Each package is declared explicitly instead of as in a wildcard import

securespider commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,70 @@
+
+    /**
+     * Function to process the user's request and get the response
+     * @return
```
Remember to specify what the method is returning

securespider commented on src/main/java/duke/storage/LocalStorage.java:
```diff
@@ -0,0 +1,143 @@
+        try {
+            file.createNewFile();
+        } catch (IOException io_error) {
+            io_error.printStackTrace();
```
Might not be good for the stack trace to be printed like this. Maybe you could reformat the error message to be nicer for the user to read? Otherwise, its good that it is verbose and the user knows that there is an error there

securespider commented on src/main/java/duke/storage/LocalStorage.java:
```diff
@@ -0,0 +1,143 @@
+                            tasks.add(deadline);
+                            break;
+                        } catch (DateTimeException error) {
+                            throw new InvalidArgumentException("Wrong date format! Please follow the format "
```
Similar to above, since this string is used in multiple lines for the same exception, might be good to store it in the exception itself

securespider reviewed (commented):
Good job!
Minor changes to the code can be done to make it look better and clearer for other developers.

Jnjy committed:
Add: Enums for duke

Jnjy committed:
Merge branch 'master' of github.com:Jnjy/ip

Jnjy committed:
Fix: Null exception for parser

Jnjy committed:
Add: include time in datetime formatting

Jnjy committed:
Edit: edit test case for addDeadline and addEvent command + minor edit on conversion from date to data string

Jnjy committed:
UI: Refactor text printing

Extracted String to print to be class variable in UI class.

This is to allow easier ammendment to the display message if needed.

the border of CLI display message is being extracted and declared as a
String variable under `border`.

Jnjy committed:
A-CodeQualiy: Extract input validation check

The original version caused clutter of if condition in the execute
function resulting in
* poor readability
* unneccessary repeats

New class `InputValidation` is created to handle all validation and
cleaning of the request from the user before parsing it to add command.

It is done this way as three commands, respectively `mark`, `unmark`
and `remove` have the same input validation check. Hence, to follow the DRY
principle, it is extracted out so when needed can simply call the
designated function `normaliseIndexRequest`. Also, it improves the
cleanliness of code.

Jnjy committed:
A-Streams: add streams to `find` method

Previous version uses for loop as list may get longer and for loop might
have a longer run time.

Task list is being transformed to streams and use `stream::filter` to
filter using the keyword provided. The return result is all task with
keyword collected in an array list.

This is done this way to accomodate for further extension of the list.

Jnjy committed:
A-Assertions: add assertion to parse command check

Default case in `Parser` is not reachable since there is a prior check
to `UnknownCommandException`

Add assert to the default case to detect any error if the program
happens to reach teh default case.

Jnjy committed:
Merge pull request #1 from Jnjy/branch-A-CodeQuality

A-CodeQuality: Extract input validation check

Jnjy committed:
Conflict: resolve conflict from branch-A-CodeQuality

Jnjy committed:
Merge pull request #3 from Jnjy/branch-A-Assertions

A-Assertions: add assertion to parse command check

Jnjy committed:
Conflict: resolve conflict from master

Jnjy committed:
Merge pull request #2 from Jnjy/branch-A-Streams

A-Streams: add streams to `find` method

Jnjy committed:
Create gradle.yml

Jnjy committed:
Fix: minor fix on checkstyle error

Jnjy committed:
Fix: checkstyle error fix on JUnit test case

Jnjy committed:
Gradle: remove text-ui-test

Jnjy committed:
C-DetectDuplicates: Add static function to prevent duplicated entry

Jnjy committed:
Merge branch 'branch-C-DetectDuplicates'

Jnjy committed:
Code Quality: Minor fix on code based on the feedback by SE-bot

Jnjy committed:
More personality: give chadbot a personality and edited the dialogbox

Jnjy committed:
Add: UI png - chadbot screenshot

Jnjy committed:
User Guide: Add first draft to user guide

Jnjy committed:
Rename: UI.png -> Ui.png

Jnjy committed:
Add: jar file reference in readme

Jnjy committed:
Edit: fix typo

Jnjy committed:
Update README.md

Jnjy committed:
Fix: change classname for jar release

Jnjy committed:
Merge branch 'master' of github.com:Jnjy/ip

Jnjy committed:
Fix: Readme Table alignment

Jnjy committed:
Add: add _config.yml file for github page table rendering

Jnjy committed:
Update README.md

Jnjy committed:
Edit: Update User Guide's alignment

Jnjy committed:
Fix: Save task list upon exit

Jnjy committed:
Fix: edit code to fix checkstyle error

Jnjy committed:
Refactor: extract loadFile from text file to be in InputValidator for better abstraction

Jnjy committed:
Fix: remove bugged emoji
